model/glove/glove.go

```diff
@@ -0,0 +1,279 @@
+}
+
+// NewGloVe creates *GloVe.
+func NewGloVe(config *model.Config, iteration int, alpha float64, xmax int) *GloVe {
+	// WITHOUT WHITESPACE, UNKNOWN, ROOT
+	emptyOpt := func(c *corpus.Corpus) error { return nil }
+	c, _ := corpus.Construct(emptyOpt)
+
+	return &GloVe{
+		Config: config,
+		Corpus: c,
+
+		CofreqMap: make(CofreqMap),
+
+		iteration: iteration,
+		alpha:     alpha,
+		xmax:      xmax,
+
+		currentLearningRate: config.InitLearningRate,
+	}
+}
+
+func (g *GloVe) update(words []string) {
+	for i := 0; i < len(words); i++ {
+		g.Add(words[i])
+		for j := i - g.Config.Window; j <= i+g.Config.Window; j++ {
+			if i == j || j < 0 || j >= len(words) {
+				continue
+			}
+			g.CofreqMap.update(words[i], words[j], math.Abs(float64(i-j)))
+		}
+	}
+	return
+}
+
+// Preprocess scans the corpus once before Train to count the co-frequency between word-word.
+func (g *GloVe) Preprocess(f io.ReadSeeker) (io.ReadCloser, error) {
+	defer func() {
+		g.emb = NewEmbedding(g.Type, g.Corpus.Size(), g.Dimension)
+		g.pairs = g.CofreqMap.toList(g.Corpus)
+	}()
+
+	buffered := 0
+	line := make([]string, 0, 10000)
+
+	scanner := bufio.NewScanner(f)
+	scanner.Split(bufio.ScanWords)
+	for scanner.Scan() {
+
+		if buffered < 10000 {
+			line = append(line, scanner.Text())
+			buffered++
+			continue
+		}
+
+		if g.ToLower {
+			model.Lower(line)
+		}
+
+		g.update(line)
+		buffered = 0
+		line = line[:0]
+	}
+
+	g.update(line)
+
+	if err := scanner.Err(); err != nil && err != io.EOF {
+		return nil, errors.Wrap(err, "Unable to complete scanning")
+	}
+
+	if _, err := f.Seek(0, 0); err != nil {
+		return nil, errors.Wrap(err, "Unable to rewind file")
+	}
+	return f.(io.ReadCloser), nil
+}
+
+// Train trains a corpus. It assumes that Preprocess() has already been called.
+func (g *GloVe) Train(f io.ReadCloser) error {
+	f.Close()
+	if len(g.pairs) == 0 {
+		return errors.Errorf("Must initialize model parameters by calling Preprocess")
+	}
+
+	fmt.Printf("The size of Corpus: %v\n", g.Size())
+	fmt.Printf("The size of CofreqMap: %v\n", len(g.pairs))
+
+	numLines := len(g.pairs)
+	perThead := make([]int, g.Thread)
+	perThead[0] = 0
+	for i := 1; i < g.Thread-1; i++ {
+		perThead[i] = numLines / g.Thread
+	}
+	perThead[g.Thread-1] = numLines/g.Thread + numLines%g.Thread
+
+	sema := make(chan struct{}, g.Thread)
+	var wg sync.WaitGroup
+
+	for i := 0; i < g.iteration; i++ {
+		for j := 0; j < len(perThead)-1; j++ {
+			wg.Add(1)
+			go g.trainPerThread(i, perThead[j], perThead[j+1], &wg, sema)
+		}
+		wg.Wait()
+		fmt.Printf("%v-th iteration: %v\n", i+1, time.Now())
+	}
+	return nil
+}
+
+func (g *GloVe) trainPerThread(sid, start, end int, wg *sync.WaitGroup, sema chan struct{}) {
+	defer wg.Done()
+	sema <- struct{}{}
+
+	for i := start; i < end; i++ {
+		p := g.pairs[i]
+		g.train(p.l1, p.l2, p.f)
+	}
+	<-sema
+}
+
+func (g *GloVe) train(pind, qind int, f float64) (err error) {
+	// SGD
+	inner, _ := tensor.Inner(g.emb.P[pind], g.emb.Q[qind])
```

Comment:
bad habit! should always check for errors.

Perhaps create a utility struct somewhere that looks like this:

```
type maybe struct{
    err error
}
func (m *maybe) DoBinOp(fn func(a, b interface{})(tensor.Tensor, error), a, b tensor.Tensor) (retVal tensor.Tensor) {
    if m.err != nil {
        return nil
    }
    retVal, m.err = fn(a, b)
    return
}

func (m *maybe) Error() error { return m.err }
```

then you can do this:

```
m := new(maybe)
inner := m.Do(tensor.Inner, g.emb.p[pind], g.emb.Q[qind])
bias := m.Do(tensor.Add, g.emb.biasP[pind, g.emb.biasQ[qind])
// and so on and so forth
```

Reply:
Sure. I just wonder how to avoid to write many `if err ...`. it's beneficial info for me, thanks.